internal/hcsoci/cmd.go

```diff
@@ -0,0 +1,318 @@
+				CreateStdErrPipe: c.Stderr != nil,
+			},
+			OCIProcess: c.Spec,
+		}
+		x = lpp
+	}
+	if c.Context != nil && c.Context.Err() != nil {
+		return c.Context.Err()
+	}
+	p, err := c.Host.CreateProcess(x)
+	if err != nil {
+		return err
+	}
+	c.Process = p
+	if c.Log != nil {
+		c.Log = c.Log.WithField("pid", p.Pid())
+	}
+
+	// Start relaying process IO.
+	stdin, stdout, stderr := p.Stdio()
+	if c.Stdin != nil {
+		// Do not make stdin part of the error group because there is no way for
+		// us or the caller to reliably unblock the c.Stdin read when the
+		// process exits.
+		go func() {
+			_, err := copyAndLog(stdin, c.Stdin, c.Log, "stdin")
+			// Report the stdin copy error. If the process has exited, then the
+			// caller may never see it, but if the error was due to a failure in
+			// stdin read, then it is likely the process is still running.
+			if err != nil {
+				c.stdinErr.Store(err)
+			}
+			// Notify the process that there is no more input.
+			p.CloseStdin()
+		}()
+	}
+
+	if c.Stdout != nil {
+		c.iogrp.Go(func() error {
+			_, err := copyAndLog(c.Stdout, stdout, c.Log, "stdout")
+			return err
+		})
+	}
+
+	if c.Stderr != nil {
+		c.iogrp.Go(func() error {
+			_, err := copyAndLog(c.Stderr, stderr, c.Log, "stderr")
+			return err
+		})
+	}
+
+	if c.Context != nil {
+		go func() {
+			select {
+			case <-c.Context.Done():
+				c.Process.Kill()
+			case <-c.allDoneCh:
+			}
+		}()
+	}
+	return nil
+}
+
+// Wait waits for a command and its IO to complete and closes the underlying
+// process. It can only be called once. It returns an ExitError if the command
+// runs and returns a non-zero exit code.
+func (c *Cmd) Wait() error {
+	err := c.Process.Wait()
+	if err != nil && c.Log != nil {
+		c.Log.WithError(err).Warn("process wait failed")
+	}
+	state := &ExitState{}
+	code, err := c.Process.ExitCode()
+	if err == nil {
+		state.exited = true
+		state.code = code
+	}
+	// Terminate the IO if the copy does not complete in the requested time.
+	if c.CopyAfterExitTimeout != 0 {
+		go func() {
+			t := time.NewTimer(c.CopyAfterExitTimeout)
+			defer t.Stop()
+			select {
+			case <-c.allDoneCh:
+			case <-t.C:
+				// Close the process to cancel any reads to stdout or stderr.
+				c.Process.Close()
+				if c.Log != nil {
+					c.Log.Warn("timed out waiting for stdio relay")
+				}
+			}
+		}()
+	}
+	ioerr := c.iogrp.Wait()
+	if err == nil {
+		err = ioerr
+		if err == nil {
+			err, _ = c.stdinErr.Load().(error)
+		}
+	}
+	close(c.allDoneCh)
+	c.Process.Close()
+	c.ExitState = state
+	if err != nil {
+		return err
+	}
+	if state.exited && state.code != 0 {
+		return &ExitError{state}
+	}
+	return nil
+}
+
+// Run is equivalent to Start followed by Wait.
+func (c *Cmd) Run() error {
+	err := c.Start()
+	if err != nil {
+		return err
+	}
+	return c.Wait()
+}
+
+// Output runs a command via Run and collects its stdout into a buffer,
+// which it returns.
+func (c *Cmd) Output() ([]byte, error) {
```

Comment:
Should we add CombinedOutput as well? Seems like we often want Stderr as well

Reply:
We could, but I didn't find any place where we wanted it.

We also have the problem today that the GCS protocol does not let us establish that stdout == stderr (i.e. that we should only use a single hvsock connection to back both of them), so the interleaving will be unpredictable.

Reply:
Ah fair point.

Reply:
Having stderr is useful for debugging failures (I've had them in the docker lcow graphdriver in the past when running UVM commands).

Reply:
You can always capture stderr yourself by passing a bytes.Buffer to Stderr.